File: experimental/remote-shell/agent/protocol.go
package main

import "encoding/binary"

// Remote-shell tunnel wire protocol. Must stay byte-for-byte compatible with
// gateway/src/protocol.ts.
//
//   byte 0      : frame type
//   bytes 1..4  : stream id (uint32, big-endian; 0 = control)
//   bytes 5..N  : payload (raw PTY bytes, or JSON for control frames)

const frameHeader = 5

const (
	// control (stream id 0)
	fRegister    = 0x01 // agent -> gw   JSON { agentId }
	fRegisterAck = 0x02 // gw -> agent   JSON { ok }

	// per PTY session (stream id > 0)
	fOpen   = 0x10 // gw -> agent   JSON { cols, rows }
	fData   = 0x11 // either        raw bytes (stdin from browser / stdout to browser)
	fResize = 0x12 // gw -> agent   JSON { cols, rows }
	fClose  = 0x13 // either        session teardown
	fExit   = 0x14 // agent -> gw   JSON { code }
)

func encodeFrame(ftype byte, streamID uint32, payload []byte) []byte {
	buf := make([]byte, frameHeader+len(payload))
	buf[0] = ftype
	binary.BigEndian.PutUint32(buf[1:5], streamID)
	copy(buf[frameHeader:], payload)
	return buf
}

func decodeFrame(data []byte) (ftype byte, streamID uint32, payload []byte) {
	if len(data) < frameHeader {
		return 0, 0, nil
	}
	return data[0], binary.BigEndian.Uint32(data[1:5]), data[frameHeader:]
}

File: experimental/remote-shell/gateway/src/protocol.ts
// Remote-shell tunnel wire protocol (agent <-> gateway).
// Must stay byte-for-byte compatible with agent/protocol.go.
//
//   byte 0      : frame type
//   bytes 1..4  : stream id (uint32 big-endian; 0 = control)
//   bytes 5..N  : payload (raw PTY bytes, or JSON for control frames)

export const FRAME_HEADER = 5;

export enum Frame {
  REGISTER = 0x01, // agent -> gw   JSON { agentId }
  REGISTER_ACK = 0x02, // gw -> agent   JSON { ok }
  OPEN = 0x10, // gw -> agent   JSON { cols, rows }
  DATA = 0x11, // either        raw PTY bytes
  RESIZE = 0x12, // gw -> agent   JSON { cols, rows }
  CLOSE = 0x13, // either        session teardown
  EXIT = 0x14, // agent -> gw    JSON { code }
}

export function encodeFrame(
  type: Frame,
  streamId: number,
  payload: Buffer | Uint8Array = Buffer.alloc(0),
): Buffer {
  const body = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
  const buf = Buffer.allocUnsafe(FRAME_HEADER + body.length);
  buf.writeUInt8(type, 0);
  buf.writeUInt32BE(streamId >>> 0, 1);
  body.copy(buf, FRAME_HEADER);
  return buf;
}

export function decodeFrame(data: Buffer): {
  type: Frame;
  streamId: number;
  payload: Buffer;
} {
  return {
    type: data.readUInt8(0) as Frame,
    streamId: data.readUInt32BE(1),
    payload: data.subarray(FRAME_HEADER),
  };
}

export function encodeJson(type: Frame, streamId: number, obj: unknown): Buffer {
  return encodeFrame(type, streamId, Buffer.from(JSON.stringify(obj), "utf8"));
}
